packages/sources/universal-v3/src/utils/LambdaController.ts

```diff
@@ -0,0 +1,348 @@
+import AWS, { Lambda } from 'aws-sdk'
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+import type { AdapterConfig } from '@chainlink/external-adapter-framework/config'
+import { customSettings } from '..'
+import type { SandboxOutput } from './buildResponse'
+
+interface SandboxRequestData {
+  source: string
+  requestId: string
+  secrets: Record<string, unknown>
+  args?: string[]
+  numAllowedQueries?: number
+  maxResponseBytes?: number
+}
+
+type TimeLastUsed = number
+
+interface DeployedLambdaFunctions {
+  timeLastUsed: Record<string, TimeLastUsed>
+  count: number
+}
+
+export class LambdaController {
+  private deployedLambdaFunctions?: DeployedLambdaFunctions
+  private initalized = false
+  private lambda: AWS.Lambda
+  private logger = makeLogger('LambdaController')
+
+  constructor(private config: AdapterConfig<typeof customSettings>) {
+    this.lambda = new AWS.Lambda({ apiVersion: '2015-03-31' })
+  }
+
+  public initalize = async (): Promise<void> => {
+    if (this.initalized) {
+      throw Error('LambdaController already initalized')
+    }
+    AWS.config.update({
+      region: process.env['AWS_REGION'] as string,
+      credentials: {
+        accessKeyId: this.config.AWS_ACCESS_KEY_ID,
+        secretAccessKey: this.config.AWS_SECRET_ACCESS_KEY,
+      },
+    })
+    this.deployedLambdaFunctions = await this.getDeployedLambdaFunctions()
+    this.logger.debug(
+      `${this.deployedLambdaFunctions?.count} Deployed Lambda functions: ${Object.keys(
+        this.deployedLambdaFunctions.timeLastUsed,
+      )}`,
+    )
+    this.initalized = true
+  }
+
+  // This background tasks monitors all deployed Lambda functions.
+  // It is used by the backgroundExecuter of the EAv3 framework
+  // When LAMBDA_MAX_DEPLOYED_FUNCTIONS is reached, the least recently used function will be deleted
+  public lambdaPruner = async (): Promise<number> => {
+    if (!this.deployedLambdaFunctions) {
+      throw Error('LambdaController class not initalized')
+    }
+
+    if (this.deployedLambdaFunctions.count < this.config.LAMBDA_MAX_DEPLOYED_FUNCTIONS) {
+      return this.config.LAMBDA_PRUNER_LOOP_WAIT_MS
+    }
+
+    const leastRecentlyUsedFunction = await this.getLeastRecentlyUsedFunction()
+
+    try {
+      await this.deleteLambdaFunction(leastRecentlyUsedFunction)
+    } catch (error) {
+      this.logger.error(`Error deleteing lambda function ${leastRecentlyUsedFunction}: ${error}`)
+      return 1
+    }
+
+    delete this.deployedLambdaFunctions.timeLastUsed[leastRecentlyUsedFunction]
```

Comment:
Does this only clean it up in the EA state?
What about cleaning it up from AWS Lambda?

Reply:
On line 68, we try to delete the Lambda function.  If the delete succeeds (ie: no error is caught), we update the EA state by deleting the entry in `deployedLambdaFunctions.timeLastUsed`

Reply:
Should this be in the try block then?

Like if we can't delete it from AWS, then it still exists. Maybe re-try the deletion?

Reply:
Oh, there is a return statement in the catch block, but yeah I could improve this for readability.